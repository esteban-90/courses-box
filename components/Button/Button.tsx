import { MouseEvent } from 'react'
import styled from '@emotion/styled'
import { boxShadow, transition, getColors, Color } from '@/components/styles'

export type ButtonProps = {
  children: string
  color?: Color
  onClick: (event: MouseEvent<HTMLButtonElement>) => void
}

export const Button = styled.button<ButtonProps>`
  all: unset;
  display: flex;
  justify-self: center;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  font-size: 1.6rem;
  width: 15rem;
  height: 4rem;
  border-radius: 1rem;

  &:hover {
    opacity: 0.9;
  }

  &:active {
    ${({ theme }) => boxShadow(theme.components.shadow1, theme.components.shadow2, true)}
  }

  ${transition()}
  ${({ color, theme }) => getColors(theme, color)}
  ${({ theme }) => boxShadow(theme.components.shadow1, theme.components.shadow2)}
`

Button.defaultProps = { color: 'primary' }
