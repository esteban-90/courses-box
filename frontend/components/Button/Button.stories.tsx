import type { ComponentStoryObj, ComponentMeta } from '@storybook/react'
import { screen, userEvent } from '@storybook/testing-library'
import { Button } from '@/components/Button'

export default {
  title: 'Controls/Button',
  component: Button,
  args: {
    children: 'Button',
  },
} as ComponentMeta<typeof Button>

export const PrimaryButton: ComponentStoryObj<typeof Button> = {
  async play() {
    await userEvent.click(screen.getByRole('button'))
  },

  args: {
    action: 'primary',
  },
}

export const SecondaryButton: ComponentStoryObj<typeof Button> = {
  ...PrimaryButton,
  args: {
    action: 'secondary',
  },
}

export const WarningButton: ComponentStoryObj<typeof Button> = {
  ...PrimaryButton,
  args: {
    action: 'warning',
  },
}

export const DangerButton: ComponentStoryObj<typeof Button> = {
  ...PrimaryButton,
  args: {
    action: 'danger',
  },
}
