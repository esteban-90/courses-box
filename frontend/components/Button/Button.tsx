import { FC } from 'react'
import styled from '@emotion/styled'
import { makeShadow } from '@/styles/helpers'
import { ButtonProps, DefinedButton } from '@/types'

export const Button = styled.button<ButtonProps>`
  all: unset;
  display: flex;
  justify-self: center;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  font-size: 1.6rem;
  width: 15rem;
  height: 4rem;
  border-radius: 1rem;
  transition: all 0.4s ease;

  color: ${({ color, theme }) => {
    const $font = color === 'secondary' ? 'regular' : color === 'warning' ? 'warning' : 'button'
    return theme.font[$font]
  }};

  background-color: ${({ color = 'primary', theme }) => {
    const $color = color === 'secondary' ? 'transparent' : theme.components[color]
    return $color
  }};

  box-shadow: ${({ theme }) => {
    const { shadow1, shadow2 } = theme.components
    const $shadow = makeShadow(shadow1, shadow2)
    return $shadow
  }};

  &:active {
    box-shadow: ${({ theme }) => {
      const { shadow1, shadow2 } = theme.components
      const $shadow = makeShadow(shadow1, shadow2, true)
      return $shadow
    }};
  }

  &:hover {
    opacity: 0.9;
  }
`

export const PrimaryButton: FC<DefinedButton> = (props): JSX.Element => <Button color='primary' {...props} />
export const SecondaryButton: FC<DefinedButton> = (props): JSX.Element => <Button color='secondary' {...props} />
export const DangerButton: FC<DefinedButton> = (props): JSX.Element => <Button color='danger' {...props} />
export const WarningButton: FC<DefinedButton> = (props): JSX.Element => <Button color='warning' {...props} />
