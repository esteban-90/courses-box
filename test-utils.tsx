import { FC, ReactElement, ReactNode } from 'react'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'
import { Themes } from '@/styles/themes'

const Wrapper: FC<{ children: ReactNode }> = ({ children }): JSX.Element => {
  return <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
}

const customRender = (ui: ReactElement, options?: RenderOptions): RenderResult => {
  return render(ui, { wrapper: Wrapper, ...options })
}

export * from '@testing-library/react'
export { customRender as render }

// 14
