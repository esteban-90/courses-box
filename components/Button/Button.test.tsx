import userEvent from '@testing-library/user-event'
import { render, screen } from '@/test-utils'
import { Button } from './Button'

describe('Button test cases', () => {
  it('Render check', () => {
    const onClick = jest.fn()
    const { asFragment } = render(<Button onClick={onClick}>Click here</Button>)

    expect(asFragment()).toMatchSnapshot()
  })

  it('Check onClick callback', async () => {
    const onClick = jest.fn()

    render(<Button onClick={onClick}>Click here</Button>)
    await userEvent.click(screen.getByRole('button'))
    expect(onClick).toHaveBeenCalled()
  })
})
