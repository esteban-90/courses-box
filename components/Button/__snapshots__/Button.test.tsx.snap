// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button test cases Render check 1`] = `
<DocumentFragment>
  <button
    class="css-sx6prm"
  >
    Click here
  </button>
</DocumentFragment>
`;
